File: playground/knowledge/cs/programming-languages/languages/c/practice/problems/movie_scheduling/mergeSort.js
function divide(arr) {
  let lhs = arr.slice(0, arr.length / 2);
  let rhs = arr.slice(arr.length / 2);

  return [lhs, rhs];
}

function conquer(lhs, rhs) {
  let leftPointer = 0;
  let rightPointer = 0;
  const result = [];

  while (leftPointer <= lhs.length - 1 || rightPointer <= rhs.length - 1) {
    const leftItem = lhs[leftPointer];
    const rightItem = rhs[rightPointer];

    if (leftItem == undefined || leftItem > rightItem) {
      result.push(rightItem);
      rightPointer++;
    } else {
      result.push(leftItem);
      leftPointer++;
    }
  }

  return result;
}

let merge = false;

function mergeSort(arr) {
  if (arr.length === 1) {
    return arr;
  }

  const [lhs, rhs] = divide(arr);
  const sortedLhs = mergeSort(lhs);
  const sortedRhs = mergeSort(rhs);

  return conquer(sortedLhs, sortedRhs);
}

const arr = [];

for (let i = 0; i < 100; i++) {
  const rand = Math.floor(Math.random() * 100);
  arr.push(rand);
}

console.log(arr);

const sorted = mergeSort(arr);
console.log(sorted);

File: playground/knowledge/cs/programming-languages/languages/c/practice/problems/movie_scheduling/merge_sort.c
#include <stdlib.h>
#include <stdio.h>

struct Tuple
{
    int *lhs;
    int *rhs;
} typedef tuple;

int *divide(int *arr, int size)
{
    int *lhs = malloc(size / 2 * sizeof(int));
    int *rhs = malloc(size / 2 * sizeof(int));
    int *result = malloc(sizeof(char));

    for (int i = 0; i < 5; i++)
    {
        result[i] = i + 1;
    }

    return result;
}

int main()
{
    int numbers[] = {4, 3, 2, 1};
    int *result = divide(numbers, sizeof(numbers) / sizeof(numbers[0]));

    for (int i = 0; i < 10; i++)
    {
        printf("r: %d\n", result[i]);
    }

    free(result);

    return 0;
}